scanner/src/funTest/kotlin/ScannerFunTest.kt

```diff
@@ -0,0 +1,217 @@
+            }
+
+            override fun getRawResult(resultsFile: File): JsonNode {
+                return yamlMapper.nullNode()
+            }
+
+            override val version: String
+                get() = "1.0.0"
+        }
+}
+
+/** Name of the file holding the expected scan result. */
+private const val EXPECTED_RESULT_FILE_NAME = "test-scanner-expected-output-for-analyzer-result.yml"
+
+/** Path to test files. */
+private val assetsDir = File("src/funTest/assets")
+
+/**
+ * Regular expression to match temporary download paths in a scan result. As these paths change on each test run,
+ * they have to be dealt with when comparing results.
+ */
+private val DOWNLOAD_PATH_REGEX = Regex("-download(\\d+)/")
+
+/**
+ * Regular expression to match the *variables* element in a scan result. Variables depend on the current
+ * environment; so they have to be deleted when comparing results.
+ */
+private val VARIABLES_REGEX = Regex("variables:.*?tool_versions:", RegexOption.DOT_MATCHES_ALL)
+
+/** Some labels to assign to the test result. */
+private val labels = mapOf(
+    "test" to "true",
+    "scan" to "full"
+)
+
+/** Stores a result object with the expected scanner result to compare to in tests. */
+private var expectedScannerResult = constructExpectedResult()
+
+/**
+ * Patch the given [scan result][result] by replacing variables part, so that it can be compared.
+ */
+private fun patchResult(result: String): String =
+    stripTempPaths(stripVariables(result))
+
+/**
+ * Replace temporary path names in the given [result].
+ */
+private fun stripTempPaths(result: String): String =
+    result.replace(DOWNLOAD_PATH_REGEX, "/")
+
+/**
+ * Remove the section with environment variables from the given [result].
+ */
+private fun stripVariables(result: String): String =
+    result.replace(VARIABLES_REGEX, "tool_versions:")
+
+/**
+ * Read the file with the expected scanner result, replace variable parts, and convert it to an [OrtResult] object.
+ */
+private fun constructExpectedResult(): OrtResult {
+    val expectedResultYaml = patchResult(
+        patchExpectedResult(
+            assetsDir.resolve("test-scanner-expected-output-for-analyzer-result.yml"),
+            revision = "HEAD"
+        )
+    )
+
+    return deserializeResult(expectedResultYaml)
+}
+
+/**
+ * Compares the [result] specified against the expected scanner result. Note that comparison needs to be done on
+ * object level, as the raw YAML representation may differ. The [result] passed in needs to be patched first to
+ * deal with variable parts. Also, access statistics cannot be compared directly.
+ */
+private fun checkResult(result: OrtResult) {
+    val patchedYaml = patchResult(
+        patchActualResult(
+            yamlMapper.writeValueAsString(result),
+            patchDownloadTime = true,
+            patchStartAndEndTime = true
+        )
+    )
+    val patchedResult = deserializeResult(patchedYaml)
```

Comment:
What you describe is certainly a good approach, and `ScannerFunTest` actually compares the result objects rather than doing a text-based comparison of the result files.

However, in this special case I wanted to make sure that the changed serialization logic does not break anything. The elements of the result are now partly written in a different order, and the test should verify that the deserialized objects are still equivalent. Therefore, I use an original result as reference.